### PR TITLE
allow ref on locals, globals, and statics

### PR DESCRIPTION
**Pull request by WalterBright**
Implementation of the proposal discussed in the n.g.:

https://www.digitalmars.com/d/archives/digitalmars/dip/development/First_Draft_ref_For_Variable_Declarations_74.html

### CODE REVIEW

File: compiler/src/dmd/dsymbolsem.d [hunk 1]
```diff
@@ -1001,9 +1001,9 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
             }
         }
 
-        if ((dsym.storage_class & (STC.ref_ | STC.parameter | STC.foreach_ | STC.temp | STC.result)) == STC.ref_ && dsym.ident != Id.This)
+        if ((dsym.storage_class & (STC.ref_ | STC.field)) == (STC.ref_ | STC.field) && dsym.ident != Id.This)
         {
-            .error(dsym.loc, "%s `%s` - only parameters, functions and `foreach` declarations can be `ref`", dsym.kind, dsym.toPrettyChars);
+            .error(dsym.loc, "%s `%s` - field declarations can be `ref`", dsym.kind, dsym.toPrettyChars);
```
Comment (thewilsonator) [inline compiler/src/dmd/dsymbolsem.d:1006]: need to update `fail_compilation/diag9679.d`. Also the error message looks like it should be improved

**Review by thewilsonator — COMMENTED** [review]
are these such `ref`s implicitly `scope`? if not should they be? Also this needs a changelog entry

**Comment by WalterBright** [comment]
> are these such refs implicitly scope`?

Yes.

**Comment by thewilsonator** [comment]
Good to hear. Please add a `fail_compilation` (e.g. to `diag9679.d`) that tests that.

**Comment by WalterBright** [comment]
> Please add a fail_compilation (e.g. to diag9679.d) that tests that.

done.

File: compiler/test/fail_compilation/diag9679.d [hunk 2]
```diff
@@ -1,13 +1,23 @@
 /*
 TEST_OUTPUT:
 ---
-fail_compilation/diag9679.d(11): Error: variable `diag9679.main.n` - only parameters, functions and `foreach` declarations can be `ref`
-fail_compilation/diag9679.d(12): Error: variable `diag9679.main.n` - storage class `auto` has no effect if type is not inferred, did you mean `scope`?
+fail_compilation/diag9679.d(14): Error: variable `diag9679.main.n` - storage class `auto` has no effect if type is not inferred, did you mean `scope`?
+fail_compilation/diag9679.d(15): Error: variable `diag9679.main.S.a` - field declarations cannot be `ref`
+fail_compilation/diag9679.d(22): Error: returning `r` escapes a reference to local variable `r`
```
Comment (thewilsonator) [inline compiler/test/fail_compilation/diag9679.d:6]: `r` escapes reference to `r`? 
`r` is a reference to `i`, not to itself.

**Comment by WalterBright** [comment]
Cybershadow/DAutoTest gives no indication why it failed.

**Comment by thewilsonator** [comment]
As always , scroll to the bottom and work your way back up:
```
object.Exception@../tools/changed.d(256): Changelog entries should consist of one title line, a blank separator line, and a description.
```

**Comment by WalterBright** [comment]
@thewilsonator I did scroll through it, but did not see that message, lost in the thousands of lines of output. Thanks for finding it for me!

File: changelog/dmd.reflocal.dd [hunk 3]
```diff
@@ -0,0 +1,12 @@
+`ref` can now be applied to local, static, extern, and global variables
+
+For example, one can now write:
+```
+void main()
+{
+    int i;
+    ref int r = i;
+    r = 3;
+    assert(i == 3);
```
Comment (ntrel) [inline changelog/dmd.reflocal.dd:10]: That could be achieved with `alias r = i;`, so I suggest:
```suggestion
struct S { int a; }

void main()
{
    S s;
    ref int r = s.a;
    r = 3;
    assert(s.a == 3);
```

File: changelog/dmd.reflocal.dd [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+}
+```
```
Comment (ntrel) [inline changelog/dmd.reflocal.dd:12]: I think the changelog should also mention any restrictions on local ref variables.

**Review by ntrel — COMMENTED** [review]
What about `(ref i) { ref r = i; ... }` - is the declaration of `r` an error?

**Comment by WalterBright** [comment]
@ntrel it's the same as the way `ref` is used in any other context. Same semantics.

File: compiler/test/fail_compilation/diag9679.d (hunk 2 above)
Comment (WalterBright) [reply compiler/test/fail_compilation/diag9679.d]: fixed it

File: changelog/dmd.reflocal.dd (hunk 4 above)
Comment (WalterBright) [reply changelog/dmd.reflocal.dd:14]: The more comprehensive documentation is in the DIP, it doesn't need repeating.

**Comment by WalterBright** [comment]
@ntrel it's the same behavior as:
```
void foo(ref int x);
void bar(ref int y)
{
    foo(y);
}
```

**Comment by WalterBright** [comment]
This is ready to go now.

**Review by dkorpel — APPROVED** [review]
The changelog makes it sound like the DIP was already accepted. Should we wait with merging or word is as a preview feature?

**Comment by thewilsonator** [comment]
We should wait, the DIP has not even been submitted, let alone reviewed by the wider community.

File: changelog/dmd.reflocal.dd (hunk 4 above)
Comment (ntrel) [reply changelog/dmd.reflocal.dd:14]: OK, I suggest a link to the DIP in the changelog.

**Comment by WalterBright** [comment]
> the DIP has not even been submitted, let alone reviewed by the wider community

That's what https://www.digitalmars.com/d/archives/digitalmars/dip/development/First_Draft_ref_For_Variable_Declarations_74.html is for.

File: changelog/dmd.reflocal.dd [hunk 5]
```diff
@@ -0,0 +1,14 @@
+`ref` can now be applied to local, static, extern, and global variables
+
```
Comment (ntrel) [inline changelog/dmd.reflocal.dd:2]:
```suggestion

See the [DIP](https://github.com/WalterBright/documents/blob/master/varRef.md).

```

File: changelog/dmd.reflocal.dd [hunk 6]
```diff
@@ -0,0 +1,14 @@
+`ref` can now be applied to local, static, extern, and global variables
```
Comment (ntrel) [inline changelog/dmd.reflocal.dd:1]: The DIP says:
> This DIP is for declaring local variables as ref. Not globals, statics, externs, __gshareds or fields.

Comment (WalterBright) [reply changelog/dmd.reflocal.dd:1]: the dip has been updated

**Comment by JohanEngelen** [comment]
please don't merge yet

**Comment by JohanEngelen** [comment]
This requires a lot more tests, not just error diagnostic testing.
A lot of things can go wrong in implementation, so please be extensive in testing.
Certainly all examples from the DIP should be tested, and things like:
```d
int global;

ref int foo() { return global; }

void main() {
    auto i = foo();
    i = 1;
    assert(global == 0);
}
```

**Comment by dkorpel** [comment]
Some things I found while testing:

`ref` can currently violate `immutable` by aliasing a mutable variable

```D
void main()
{
    int x = 3;
    immutable ref int xr = x; // accepted!
    assert(xr == 3);
    x++;
    assert(xr == 3); // fails, immutable `xr` changed
}
```

`ref` can bind to rvalues and escape references to the stack this way

```D
ref int escape() @safe
{
    ref int x = 4; // accepted!
    return x; // accepted!
}
```

**Comment by ntrel** [comment]
Perhaps add a `-preview` switch for this until it's stable enough? Then we can fix bugs separately as they come up.

**Comment by WalterBright** [comment]
I tried to rebase it, but it got all fouled up by the refactoring done to escape.d. Sigh.

**Comment by dkorpel** [comment]
> I tried to rebase it, but it got all fouled up by the refactoring done to escape.d. Sigh.

You can just remove all changes this PR does to escape.d, it's already accomodating `ref` variables now.

**Comment by WalterBright** [comment]
> ref can bind to rvalues and escape references to the stack this way

There's already a test for that in diag9679.d. Like I said, the refactor means I have to go back through and refigure out how to implement it.

**Comment by dkorpel** [comment]
> There's already a test for that in diag9679.d. Like I said, the refactor means I have to go back through and refigure out how to implement it.

You just have to check that the initializer is an lvalue in dsymbolsem, escape.d works fine currently no?

**Comment by WalterBright** [comment]
@JohanEngelen that example isn't part of the dip https://github.com/WalterBright/documents/blob/984374ca885e1cb10c2667cf872aebc13b4c1663/varRef.md

So I'm not sure what you're looking at?

**Comment by WalterBright** [comment]
> Perhaps add a -preview switch for this until it's stable enough?

As this is additive, and doesn't change existing code, it isn't necessary.

**Comment by JohanEngelen** [comment]
> @JohanEngelen that example isn't part of the dip https://github.com/WalterBright/documents/blob/984374ca885e1cb10c2667cf872aebc13b4c1663/varRef.md
> 
> So I'm not sure what you're looking at?

My point was to add the examples from the dip *and* add more. The example I gave is something that seems obvious but prevents bugs in the implementation where `auto` infers `ref` (it shouldn't).

**Comment by WalterBright** [comment]
@JohanEngelen the language never infers storage class, but I put your example in the test suite anyway.

I don't think it needs more, after all, `ref` is used in several other contexts and has been pretty thoroughly debugged.

**Comment by JohanEngelen** [comment]
A big reason why there are still so many bugs and regressions in the compiler is because tests are so scarce.
The functionality says: "`ref` can now be applied to local, static, extern, and global variables". Four categories of variables, half of which don't even have a single test case...

**Comment by JohanEngelen** [comment]
Yep, and as I already predicted:   https://issues.dlang.org/show_bug.cgi?id=24729

**Pull request merged**